### PR TITLE
Avoid some assuptions about using the GNU ld linker

### PR DESCRIPTION
**Pull request by rib**
This tries to better consider that Clang supports running with multiple alternative linkers such `ld`, `ld64` and `link.exe` and `lld` which itself has alternative interfaces based on each of those.

In particular; while cross compiling for Windows using Clang and passing `-fuse-ld=lld` that will result in `ld-link` being used which has a `link.exe` compatible interface.

This breaks an assumptions about the `--fatal-warnings` argument used for checking supported linker arguments; `--out-implib` isn't supported in this case and `--start-group`/`--end-group` brackets don't do anything.

These patches have enabled me to cross compile [https://github.com/glimpse-project/glimpse](), and it's subproject dependencies, for Windows, using Clang + LLD from Linux.

### CODE REVIEW

**Comment by rib** [comment]
Currently a bit unsure how to fix the unittest exception caused by the last patch:

```
Traceback (most recent call last):
  File "run_unittests.py", line 355, in test_compiler_args_class_gnuld
    self.assertEqual(l.to_native(env, copy=True), ['-Lfoodir', '-Wl,--start-group', '-lfoo', '-Wl,--end-group'])
  File "/root/mesonbuild/compilers/compilers.py", line 725, in to_native
    group_args = self.compiler.get_link_group_args(env)
  File "/root/mesonbuild/compilers/c.py", line 319, in get_link_group_args
    self.has_group_args = self.has_multi_link_arguments(group_args, env)
  File "/root/mesonbuild/compilers/c.py", line 1122, in has_multi_link_arguments
    self.has_fatal_warnings_link_arg = ((self.has_multi_link_arguments(fatal_warnings_args, env) and not
  File "/root/mesonbuild/compilers/c.py", line 1137, in has_multi_link_arguments
    return self.has_arguments(args, env, code, mode='link')
  File "/root/mesonbuild/compilers/c.py", line 1085, in has_arguments
    return self.compiles(code, env, extra_args=args, mode=mode)
  File "/root/mesonbuild/compilers/c.py", line 465, in compiles
    with self._build_wrapper(code, env, extra_args, dependencies, mode) as p:
  File "/usr/lib/python3.6/contextlib.py", line 83, in __enter__
    raise RuntimeError("generator didn't yield") from None
RuntimeError: generator didn't yield
```

If I remove the try/catch block I can see I'm getting a PermissionError exception like:
```
======================================================================
ERROR: test_compiler_args_class_gnuld (__main__.InternalTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./run_unittests.py", line 355, in test_compiler_args_class_gnuld
    self.assertEqual(l.to_native(env, copy=True), ['-Lfoodir', '-Wl,--start-group', '-lfoo', '-Wl,--end-group'])
  File "/Users/bob/src/meson/mesonbuild/compilers/compilers.py", line 725, in to_native
    group_args = self.compiler.get_link_group_args(env)
  File "/Users/bob/src/meson/mesonbuild/compilers/c.py", line 319, in get_link_group_args
    self.has_group_args = self.has_multi_link_arguments(group_args, env)
  File "/Users/bob/src/meson/mesonbuild/compilers/c.py", line 1122, in has_multi_link_arguments
    self.has_fatal_warnings_link_arg = ((self.has_multi_link_arguments(fatal_warnings_args, env) and not
  File "/Users/bob/src/meson/mesonbuild/compilers/c.py", line 1137, in has_multi_link_arguments
    return self.has_arguments(args, env, code, mode='link')
  File "/Users/bob/src/meson/mesonbuild/compilers/c.py", line 1085, in has_arguments
    return self.compiles(code, env, extra_args=args, mode=mode)
  File "/Users/bob/src/meson/mesonbuild/compilers/c.py", line 465, in compiles
    with self._build_wrapper(code, env, extra_args, dependencies, mode) as p:
  File "/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/Users/bob/src/meson/mesonbuild/compilers/compilers.py", line 1140, in compile
    p, p.stdo, p.stde = Popen_safe(commands, cwd=tmpdirname, env=os_env)
  File "/Users/bob/src/meson/mesonbuild/mesonlib.py", line 913, in Popen_safe
    return Popen_safe_legacy(args, write=write, stdout=stdout, stderr=stderr, **kwargs)
  File "/Users/bob/src/meson/mesonbuild/mesonlib.py", line 921, in Popen_safe_legacy
    stdout=stdout, stderr=stderr, **kwargs)
  File "/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 769, in __init__
    restore_signals, start_new_session)
  File "/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 1516, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: '/var/folders/cz/3mxyvqj13v71s2gf0d8syxzr0000gn/T/tmpxfvf76r0/testfile.c'
```

Since we're trying to run the source file not `gcc`. I guess there's something incomplete about the fake GnuCCompiler created in the unittest whereby it doesn't define the name of the compiler executable itself, but not sure yet how to fix that. Perhaps the unit test should be run differently now, actually with gcc instead of a fake compiler, considering that the compiler does now need to be run to check the feature is supported.

The feature does seem to work outside of the unit testing.

I'll investigate a bit further, although potentially I could drop this patch to start with since `lld-link` only prints warnings about unrecognised arguments when passing `--start-group`/`--end-group` brackets - so they aren't really fatal for my use case.

**Comment by rib** [comment]
Ok, adding:
```
         gcc = mesonbuild.compilers.GnuCCompiler([], 'fake', mesonbuild.compilers.CompilerType.GCC_STANDARD, False)
+        # Ensure we don't need to dynamically run compiler to check feature...
+        gcc.has_fatal_warnings_link_arg = True
+        gcc.has_group_args = True
+
```
works to ensure the fake compiler doesn't need to be run dynamically to check any features for this unit test

**Comment by jon-turney** [comment]
I have a branch with changes to tease apart the compiler:linker correspondence as part of the work to address #4232. That probably conflicts quite heavily with these changes...

**Comment by dcbaker** [comment]
I also have a branch teasing apart linker compiler while trying to get solaris working, since solaris has both gnu ld and their own incompatible ld. I think we really should break the assumption that the compiler and linker always come as a pair.

**Comment by XVilka** [comment]
So what had been decided with this PR?

**Comment by jpakkane** [comment]
Sorry for the delays that have been caused both by Fosdem and me having a massive flu.

Design-wise the important questions that I can think of are:

- Will we always be able to invoke the linker via the compiler or do we need to do it directly? I'd rather not have to do the latter, the linker command lines needed to link e.g. a simple C++ program on Linux directly with `ld` are humongous.
- Is it ok to require one global linker, i.e. either bfd for everything or gold for everything or lld for everything, or do we need to support many linkers at the same time (which we'd like to avoid because it adds complexity)

**Comment by rib** [comment]
I think it's probably reasonable to rely on the compiler frontend to handle some finer details of linking without having to e.g. invoke `ld` manually, but don't think it'll be reasonable to try and require one linker to rule them all.

Personally I have a project that we want to support building for Android, iOS, OSX, Windows and Linux, so basically end up needing to link with ld64, link and ld.

I guess those are the main three linker flavours that are relevent to a large number of developers that each deal with significant-enough operating system differences that you e.g. wouldn't really want to try and avoid using ld64 for OSX/iOS builds or avoid using link with .pdb support for Windows builds (unless targetting mingw).

I think it would be best to avoid assuming anything based on compiler version string checks and rather explicitly detect the linker flavour by checking for supported arguments. (e.g. I think meson probably assumes ld64 if it detects Clang on OSX).

Since it's really nice to be able to use lld now for cross compiling I think it's valuable to consider that those three flavours may come from different linker implementations so certain features should still be probed for. For Linux we could e.g. have GNU ld, Gold or lld as implementations, on Windows it could be Microsoft's link.exe or lld-link and for OSX/iOS it might be Apple's ld64 or lld-ld64. I'm not really sure if lld-ld64 is being actively developed or used currently, but notably it is possible build ld64 on Linux which is what we were using at one point for cross-compiling iOS builds from Linux.

In summary my instinct would be to have Meson be specifically aware of three linker flavours: LD, LD64 and LINK. Meson should try its best to infer which flavour is in use by explicitly testing for valid arguments instead of assuming anything based on version string parsing. For each flavour there might be a core set of assumed features but Meson should allow multiple implementations of each flavour and explicitly check for some features.

**Comment by XVilka** [comment]
> In summary my instinct would be to have Meson be specifically aware of three linker flavours: LD, LD64 and LINK. Meson should try its best to infer which flavour is in use by explicitly testing for valid arguments instead of assuming anything based on version string parsing. For each flavour there might be a core set of assumed features but Meson should allow multiple implementations of each flavour and explicitly check for some features.

Totally agree with this.

**Comment by dcbaker** [comment]
@rib, is this still a problem on 0.52.0 or master?

**Comment by rib** [comment]
Briefly trying to build the glimpse `wip/rib/next` branch with meson master on windows (WSL) I currently get this build failure:

```
meson --cross-file ../windows-sdk/meson/windows-x64-debug-cross-file.txt
The Meson build system
Version: 0.52.999
Source dir: /home/rib/glimpse
Build dir: /home/rib/glimpse/build-debug-win
Build type: cross build
Project name: glimpse
Project version: undefined
C compiler for the build machine: cc (gcc 7.4.0 "cc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0")
C linker for the build machine: cc GNU ld.bfd 2.30
C++ compiler for the build machine: c++ (gcc 7.4.0 "c++ (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0")
C++ linker for the build machine: c++ GNU ld.bfd 2.30
Traceback (most recent call last):
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 129, in run
    return options.run_func(options)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 245, in run
    app.generate()
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 159, in generate
    self._generate(env)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 176, in _generate
    intr = interpreter.Interpreter(b)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 2114, in __init__
    self.parse_project()
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 397, in parse_project
    self.evaluate_codeblock(self.ast, end=1)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 436, in evaluate_codeblock
    raise e
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 430, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 441, in evaluate_statement
    return self.function_call(cur)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 788, in function_call
    return func(node, posargs, kwargs)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 143, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 174, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 2790, in func_project
    self.add_languages(proj_langs, True)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 2851, in add_languages
    success &= self.add_languages_for(args, required, MachineChoice.HOST)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 2866, in add_languages_for
    comp = self.environment.detect_compiler_for(lang, for_machine)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/environment.py", line 1541, in detect_compiler_for
    comp = self.compiler_from_language(lang, for_machine)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/environment.py", line 1513, in compiler_from_language
    comp = self.detect_c_compiler(for_machine)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/environment.py", line 1070, in detect_c_compiler
    return self._detect_c_or_cpp_compiler('c', for_machine)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/environment.py", line 998, in _detect_c_or_cpp_compiler
    linker = self._guess_win_linker(compiler, cls, for_machine)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/environment.py", line 762, in _guess_win_linker
    p, o, e = Popen_safe(compiler + check_args)
  File "/home/rib/.local/lib/python3.6/site-packages/mesonbuild/mesonlib.py", line 1072, in Popen_safe
    stdout=stdout, stderr=stderr, **kwargs)
  File "/usr/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.6/subprocess.py", line 1364, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'ldd-8': 'ldd-8'
```

(while I do have ldd-8 installed:)
```
$ which lld-8
/usr/bin/lld-8
```

So I'm not really sure currently.

It's no longer critical for me since I'm no longer actively working on that project so I'm not sure I'll get a chance to re-investigate this soon.

Rough instructions in case someone wanted to try and reproduce the use case would be:

Under Windows WSL (Ubuntu 18.04):
```
sudo apt-get install curl     unzip     build-essential     ninja-build     clang-6.0     lld-6.0     clang-8     lld-8     python3-minimal     python3-pip     python3-setuptools     git     pkg-config     libz-dev     libpng-dev     libepoxy-dev     libfreenect-dev     libglfw3-dev     libglm-dev     libpcl-dev
git clone https://github.com/glimpse-project/glimpse --branch wip/rib/next
cd glimpse
./windows-sdk-build.py
mkdir build-win
cd build-win
meson --cross-file ../windows-sdk/meson/windows-x64-debug-cross-file.txt
```
(The windows-sdk-build.py script builds a portable Windows SDK (headers and libraries) that can be used for cross compiling Windows applications from Linux. It requires visual studio >= 2017 + windows 10 sdk be installed when running the script)

(Make sure to fetch and run while under WSL's case-sensitive filesystem)

For reference, this is an example of the generated meson cross file:
```
[binaries]
name = 'clang-8-windows-x64-debug'
c = 'clang-8'
cpp = 'clang++-8'
ar = 'llvm-ar'
ld = 'ldd-8'
strip = 'llvm-strip'

[host_machine]
system = 'windows'
cpu_family = 'x86_64'
cpu = 'x86_64'
endian = 'little'

[properties]
sdk_path = '/home/rib/glimpse/windows-sdk'

c_args = [ '-target', 'x86_64-pc-windows', '-fms-extensions', '-fms-compatibility', '-fdelayed-template-parsing', '-Wno-expansion-to-defined', '-Wno-ignored-attributes', '-isystem', '/home/rib/glimpse/windows-sdk/vc/include', '-isystem', '/home/rib/glimpse/windows-sdk/ucrt/include', '-isystem', '/home/rib/glimpse/windows-sdk/sdk/um/include', '-isystem', '/home/rib/glimpse/windows-sdk/sdk/shared/include', '-D_WINSOCK_DEPRECATED_NO_WARNINGS=1', '-DNOMINMAX=1', '-D_CRT_DECLARE_NONSTDC_NAMES=1', '-D_CRT_NONSTDC_NO_DEPRECATE=1', '-D_CRT_SECURE_NO_DEPRECATE=1', '-D_CRT_NONSTDC_NO_WARNINGS=1', '-D_CRT_SECURE_NO_WARNINGS', '-fuse-ld=lld' ]

c_link_args = [ '-target', 'x86_64-pc-windows', '-fms-compatibility', '-L/home/rib/glimpse/windows-sdk/vc/lib/x64', '-L/home/rib/glimpse/windows-sdk/ucrt/lib/x64', '-L/home/rib/glimpse/windows-sdk/sdk/um/lib/x64', '-fuse-ld=lld', '-nostdlib', '-lmsvcrtd', '-lvcruntimed', '-lucrtd', '-g' ]

cpp_args = [ '-target', 'x86_64-pc-windows', '-fms-extensions', '-fms-compatibility', '-fdelayed-template-parsing', '-Wno-expansion-to-defined', '-Wno-ignored-attributes', '-isystem', '/home/rib/glimpse/windows-sdk/vc/include', '-isystem', '/home/rib/glimpse/windows-sdk/ucrt/include', '-isystem', '/home/rib/glimpse/windows-sdk/sdk/um/include', '-isystem', '/home/rib/glimpse/windows-sdk/sdk/shared/include', '-D_WINSOCK_DEPRECATED_NO_WARNINGS=1', '-DNOMINMAX=1', '-D_CRT_DECLARE_NONSTDC_NAMES=1', '-D_CRT_NONSTDC_NO_DEPRECATE=1', '-D_CRT_SECURE_NO_DEPRECATE=1', '-D_CRT_NONSTDC_NO_WARNINGS=1', '-D_CRT_SECURE_NO_WARNINGS=1', '-fuse-ld=lld' ]

cpp_link_args = [ '-target', 'x86_64-pc-windows', '-fms-compatibility', '-L/home/rib/glimpse/windows-sdk/vc/lib/x64', '-L/home/rib/glimpse/windows-sdk/ucrt/lib/x64', '-L/home/rib/glimpse/windows-sdk/sdk/um/lib/x64', '-fuse-ld=lld', '-nostdlib', '-nostdlib++', '-lmsvcrtd', '-lvcruntimed', '-lucrtd', '-g' ]
```

**Comment by rib** [comment]
Oh that was a silly typo! `s/ldd/lld/` and then there was another fixup needed where the generated meson file referenced `llvm-ar` instead of `llvm-ar-8`.

With those fixes meson is configuring and fetching deps at least, and can let you know if it manages to build too in a bit...

**Comment by rib** [comment]
The build failed with:
```
clang-8  -o subprojects/libpng-1.6.17/libpng.dll 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/png.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngerror.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngget.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngmem.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngpread.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngread.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngrio.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngrtran.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngrutil.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngset.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngtrans.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngwio.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngwrite.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngwtran.c.obj' 'subprojects/libpng-1.6.17/9cfed1b@@png@sha/pngwutil.c.obj' -L/home/rib/glimpse/windows-sdk/vc/lib/x64 -L/home/rib/glimpse/windows-sdk/ucrt/lib/x64 -L/home/rib/glimpse/windows-sdk/sdk/um/lib/x64 -Wl,--allow-shlib-undefined -fuse-ld=lld-8 -shared -Wl,--start-group -Wl,--out-implib=subprojects/libpng-1.6.17/libpng.dll.a -target x86_64-pc-windows -fms-compatibility -fuse-ld=lld -nostdlib -lmsvcrtd -lvcruntimed -lucrtd -g subprojects/zlib-1.2.8/libz.dll.a -Wl,--end-group
clang: error: no such file or directory: 'subprojects/zlib-1.2.8/libz.dll.a'
```

Where this is the listing for ./subprojects`:
```
./subprojects/zlib-1.2.8/
./subprojects/zlib-1.2.8/90da1ba@@z@sha
./subprojects/zlib-1.2.8/90da1ba@@z@sha/adler32.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/compress.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/crc32.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/deflate.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/gzclose.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/gzlib.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/gzread.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/gzwrite.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/infback.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/inffast.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/inflate.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/inftrees.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/libz.dll.symbols
./subprojects/zlib-1.2.8/90da1ba@@z@sha/trees.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/uncompr.c.obj
./subprojects/zlib-1.2.8/90da1ba@@z@sha/zutil.c.obj
./subprojects/zlib-1.2.8/libz.dll
./subprojects/zlib-1.2.8/libz.lib
./subprojects/zlib-1.2.8/libz.pdb
```

So I guess that's going to end up being a meson linker backend issue relating to link.exe/lld-link behaviour and naming convensions around linking shared/static libraries
